datafusion/ffi/src/proto/logical_extension_codec.rs

```diff
@@ -0,0 +1,739 @@
+            PhysicalExtensionCodec::try_decode_udf(self, name, buf)
+        }
+
+        fn try_encode_udf(
+            &self,
+            node: &ScalarUDF,
+            buf: &mut Vec<u8>,
+        ) -> datafusion_common::Result<()> {
+            PhysicalExtensionCodec::try_encode_udf(self, node, buf)
+        }
+
+        fn try_decode_udaf(
+            &self,
+            name: &str,
+            buf: &[u8],
+        ) -> datafusion_common::Result<Arc<AggregateUDF>> {
+            PhysicalExtensionCodec::try_decode_udaf(self, name, buf)
+        }
+
+        fn try_encode_udaf(
+            &self,
+            node: &AggregateUDF,
+            buf: &mut Vec<u8>,
+        ) -> datafusion_common::Result<()> {
+            PhysicalExtensionCodec::try_encode_udaf(self, node, buf)
+        }
+
+        fn try_decode_udwf(
+            &self,
+            name: &str,
+            buf: &[u8],
+        ) -> datafusion_common::Result<Arc<WindowUDF>> {
+            PhysicalExtensionCodec::try_decode_udwf(self, name, buf)
+        }
+
+        fn try_encode_udwf(
+            &self,
+            node: &WindowUDF,
+            buf: &mut Vec<u8>,
+        ) -> datafusion_common::Result<()> {
+            PhysicalExtensionCodec::try_encode_udwf(self, node, buf)
+        }
+    }
+
+    #[test]
+    fn roundtrip_ffi_logical_extension_codec_table_provider(
+    ) -> datafusion_common::Result<()> {
+        let codec = Arc::new(TestExtensionCodec {});
+        let (ctx, task_ctx_provider) = crate::util::tests::test_session_and_ctx();
+
+        let mut ffi_codec =
+            FFI_LogicalExtensionCodec::new(codec, None, task_ctx_provider);
+        ffi_codec.library_marker_id = crate::mock_foreign_marker_id;
+        let foreign_codec: Arc<dyn LogicalExtensionCodec> = (&ffi_codec).into();
+
+        let table = Arc::new(create_test_table()) as Arc<dyn TableProvider>;
+        let mut bytes = Vec::new();
+        foreign_codec.try_encode_table_provider(&"my_table".into(), table, &mut bytes)?;
+
+        let returned_table = foreign_codec.try_decode_table_provider(
+            &bytes,
+            &"my_table".into(),
+            create_test_table().schema(),
+            ctx.task_ctx().as_ref(),
+        )?;
+
+        assert!(returned_table.as_any().is::<MemTable>());
+
+        Ok(())
+    }
+
+    #[test]
+    fn roundtrip_ffi_logical_extension_codec_udf() -> datafusion_common::Result<()> {
+        let codec = Arc::new(TestExtensionCodec {});
+        let (_ctx, task_ctx_provider) = crate::util::tests::test_session_and_ctx();
+
+        let mut ffi_codec =
+            FFI_LogicalExtensionCodec::new(codec, None, task_ctx_provider);
+        ffi_codec.library_marker_id = crate::mock_foreign_marker_id;
+        let foreign_codec: Arc<dyn LogicalExtensionCodec> = (&ffi_codec).into();
+
+        let udf = Arc::new(ScalarUDF::from(AbsFunc::new()));
+        let mut bytes = Vec::new();
+        foreign_codec.try_encode_udf(udf.as_ref(), &mut bytes)?;
```

Comment:
I am wondering if having a unit test that shows error handling (or expected behavior) when trying the round trip but the data we are reading is corrupted/incomplete somehow. That was at least it is clear for the consumers what the intent would be